// react-native-reanimated/lib/commonjs/reanimated2/core.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.checkPluginState = void 0;
exports.configureLayoutAnimations = configureLayoutAnimations;
exports.configureProps = configureProps;
exports.enableLayoutAnimations = enableLayoutAnimations;
Object.defineProperty(exports, "getTimestamp", {
  enumerable: true,
  get: function () {
    return _time.getTimestamp;
  }
});
exports.getViewProp = getViewProp;
exports.makeShareable = exports.makeRemote = exports.makeMutable = exports.isConfiguredCheck = exports.isConfigured = void 0;
exports.registerEventHandler = registerEventHandler;
exports.registerSensor = registerSensor;
Object.defineProperty(exports, "runOnJS", {
  enumerable: true,
  get: function () {
    return _threads.runOnJS;
  }
});
Object.defineProperty(exports, "runOnUI", {
  enumerable: true,
  get: function () {
    return _threads.runOnUI;
  }
});
exports.startMapper = void 0;
Object.defineProperty(exports, "stopMapper", {
  enumerable: true,
  get: function () {
    return _mappers.stopMapper;
  }
});
exports.subscribeForKeyboardEvents = subscribeForKeyboardEvents;
exports.unregisterEventHandler = unregisterEventHandler;
exports.unregisterSensor = unregisterSensor;
exports.unsubscribeFromKeyboardEvents = unsubscribeFromKeyboardEvents;

var _NativeReanimated = _interopRequireDefault(require("./NativeReanimated"));

var _PlatformChecker = require("./PlatformChecker");

var _shareables = require("./shareables");

var _threads = require("./threads");

var _mappers = require("./mappers");

var _mutables = require("./mutables");

var _time = require("./time");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* global _setGlobalConsole */
if (global._setGlobalConsole === undefined) {
  // it can happen when Reanimated plugin wasn't added, but the user uses the only API from version 1
  global._setGlobalConsole = () => {// noop
  };
}

const testWorklet = () => {
  'worklet';
};

const throwUninitializedReanimatedException = () => {
  throw new Error("Failed to initialize react-native-reanimated library, make sure you followed installation steps here: https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation/ \n1) Make sure reanimated's babel plugin is installed in your babel.config.js (you should have 'react-native-reanimated/plugin' listed there - also see the above link for details) \n2) Make sure you reset build cache after updating the config, run: yarn start --reset-cache");
};

const checkPluginState = function () {
  let throwError = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

  if (!testWorklet.__workletHash && !(0, _PlatformChecker.shouldBeUseWeb)()) {
    if (throwError) {
      throwUninitializedReanimatedException();
    }

    return false;
  }

  return true;
};

exports.checkPluginState = checkPluginState;

const isConfigured = function () {
  let throwError = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  return checkPluginState(throwError);
};

exports.isConfigured = isConfigured;

const isConfiguredCheck = () => {
  checkPluginState(true);
};

exports.isConfiguredCheck = isConfiguredCheck;
const configurationCheckWrapper = __DEV__ ? fn => {
  return function () {
    isConfigured(true);
    return fn(...arguments);
  };
} : fn => fn;
const startMapper = __DEV__ ? configurationCheckWrapper(_mappers.startMapper) : _mappers.startMapper;
exports.startMapper = startMapper;
const makeShareable = __DEV__ ? configurationCheckWrapper(_shareables.makeShareable) : _shareables.makeShareable;
exports.makeShareable = makeShareable;
const makeMutable = __DEV__ ? configurationCheckWrapper(_mutables.makeMutable) : _mutables.makeMutable;
exports.makeMutable = makeMutable;
const makeRemote = __DEV__ ? configurationCheckWrapper(_mutables.makeRemote) : _mutables.makeRemote;
exports.makeRemote = makeRemote;
global._WORKLET = false;

global._log = function (s) {
  console.log(s);
};

function getViewProp(viewTag, propName) {
  if (global._IS_FABRIC) {
    throw new Error('[react-native-reanimated] `getViewProp` is not supported on Fabric yet');
  }

  return new Promise((resolve, reject) => {
    return _NativeReanimated.default.getViewProp(viewTag, propName, result => {
      if (typeof result === 'string' && result.substr(0, 6) === 'error:') {
        reject(result);
      } else {
        resolve(result);
      }
    });
  });
}

function valueUnpacker(objectToUnpack, category) {
  'worklet';

  let workletsCache = global.__workletsCache;
  let handleCache = global.__handleCache;

  if (workletsCache === undefined) {
    // init
    workletsCache = global.__workletsCache = new Map();
    handleCache = global.__handleCache = new WeakMap();
  }

  if (objectToUnpack.__workletHash) {
    let workletFun = workletsCache.get(objectToUnpack.__workletHash);

    if (workletFun === undefined) {
      // eslint-disable-next-line no-eval
      workletFun = eval('(' + objectToUnpack.asString + '\n)');
      workletsCache.set(objectToUnpack.__workletHash, workletFun);
    }

    const functionInstance = workletFun.bind(objectToUnpack);
    objectToUnpack._recur = functionInstance;
    return functionInstance;
  } else if (objectToUnpack.__init) {
    let value = handleCache.get(objectToUnpack);

    if (value === undefined) {
      value = objectToUnpack.__init();
      handleCache.set(objectToUnpack, value);
    }

    return value;
  } else if (category === 'RemoteFunction') {
    const fun = () => {
      throw new Error(`Tried to synchronously call a non-worklet function on the UI thread.

Possible solutions are:
  a) If you want to synchronously execute this method, mark it as a worklet
  b) If you want to execute this function on the JS thread, wrap it using \`runOnJS\``);
    };

    fun.__remoteFunction = objectToUnpack;
    return fun;
  } else {
    throw new Error('data type not recognized by unpack method');
  }
}

function registerEventHandler(eventHash, eventHandler) {
  return _NativeReanimated.default.registerEventHandler(eventHash, (0, _shareables.makeShareableCloneRecursive)(eventHandler));
}

function unregisterEventHandler(id) {
  return _NativeReanimated.default.unregisterEventHandler(id);
}

function subscribeForKeyboardEvents(eventHandler) {
  return _NativeReanimated.default.subscribeForKeyboardEvents((0, _shareables.makeShareableCloneRecursive)(eventHandler));
}

function unsubscribeFromKeyboardEvents(listenerId) {
  return _NativeReanimated.default.unsubscribeFromKeyboardEvents(listenerId);
}

function registerSensor(sensorType, interval, eventHandler) {
  return _NativeReanimated.default.registerSensor(sensorType, interval, (0, _shareables.makeShareableCloneRecursive)(eventHandler));
}

function unregisterSensor(listenerId) {
  return _NativeReanimated.default.unregisterSensor(listenerId);
}

_NativeReanimated.default.installCoreFunctions(valueUnpacker);

if (!(0, _PlatformChecker.isWeb)() && isConfigured()) {
  const capturableConsole = console;
  (0, _threads.runOnUI)(() => {
    'worklet';

    const console = {
      debug: (0, _threads.runOnJS)(capturableConsole.debug),
      log: (0, _threads.runOnJS)(capturableConsole.log),
      warn: (0, _threads.runOnJS)(capturableConsole.warn),
      error: (0, _threads.runOnJS)(capturableConsole.error),
      info: (0, _threads.runOnJS)(capturableConsole.info)
    };

    _setGlobalConsole(console);
  })();
}

let featuresConfig = {
  enableLayoutAnimations: false,
  setByUser: false
};

function enableLayoutAnimations(flag) {
  let isCallByUser = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

  if (isCallByUser) {
    featuresConfig = {
      enableLayoutAnimations: flag,
      setByUser: true
    };

    _NativeReanimated.default.enableLayoutAnimations(flag);
  } else if (!featuresConfig.setByUser && featuresConfig.enableLayoutAnimations !== flag) {
    featuresConfig.enableLayoutAnimations = flag;

    _NativeReanimated.default.enableLayoutAnimations(flag);
  }
}

function configureLayoutAnimations(viewTag, type, config) {
  _NativeReanimated.default.configureLayoutAnimation(viewTag, type, (0, _shareables.makeShareableCloneRecursive)(config));
}

function configureProps(uiProps, nativeProps) {
  if (!(0, _PlatformChecker.nativeShouldBeMock)()) {
    _NativeReanimated.default.configureProps(uiProps, nativeProps);
  }
}
