// react-native-reanimated/lib/commonjs/reanimated2/hook/useAnimatedSensor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SensorType = void 0;
exports.useAnimatedSensor = useAnimatedSensor;

var _react = require("react");

var _core = require("../core");

let SensorType;
exports.SensorType = SensorType;

(function (SensorType) {
  SensorType[SensorType["ACCELEROMETER"] = 1] = "ACCELEROMETER";
  SensorType[SensorType["GYROSCOPE"] = 2] = "GYROSCOPE";
  SensorType[SensorType["GRAVITY"] = 3] = "GRAVITY";
  SensorType[SensorType["MAGNETIC_FIELD"] = 4] = "MAGNETIC_FIELD";
  SensorType[SensorType["ROTATION"] = 5] = "ROTATION";
})(SensorType || (exports.SensorType = SensorType = {}));

function initSensorData(sensorType) {
  if (sensorType === SensorType.ROTATION) {
    return (0, _core.makeMutable)({
      qw: 0,
      qx: 0,
      qy: 0,
      qz: 0,
      yaw: 0,
      pitch: 0,
      roll: 0
    });
  } else {
    return (0, _core.makeMutable)({
      x: 0,
      y: 0,
      z: 0
    });
  }
}

function useAnimatedSensor(sensorType, userConfig) {
  const ref = (0, _react.useRef)({
    sensor: initSensorData(sensorType),
    unregister: () => {// NOOP
    },
    isAvailable: false,
    config: {
      interval: 0
    }
  });
  (0, _react.useEffect)(() => {
    ref.current.config = {
      interval: 'auto',
      ...userConfig
    };
    const sensorData = ref.current.sensor;
    const id = (0, _core.registerSensor)(sensorType, ref.current.config.interval === 'auto' ? -1 : ref.current.config.interval, data => {
      'worklet';

      sensorData.value = data;
    });

    if (id !== -1) {
      // if sensor is available
      ref.current.unregister = () => (0, _core.unregisterSensor)(id);

      ref.current.isAvailable = true;
    } else {
      // if sensor is unavailable
      ref.current.unregister = () => {// NOOP
      };

      ref.current.isAvailable = false;
    }

    return () => {
      ref.current.unregister();
    };
  }, [sensorType, userConfig]);
  return ref.current;
}
