// react-native-reanimated/lib/module/reanimated2/hook/useAnimatedSensor.js
import { useEffect, useRef } from 'react';
import { makeMutable, registerSensor, unregisterSensor } from '../core';
export let SensorType;

(function (SensorType) {
  SensorType[SensorType["ACCELEROMETER"] = 1] = "ACCELEROMETER";
  SensorType[SensorType["GYROSCOPE"] = 2] = "GYROSCOPE";
  SensorType[SensorType["GRAVITY"] = 3] = "GRAVITY";
  SensorType[SensorType["MAGNETIC_FIELD"] = 4] = "MAGNETIC_FIELD";
  SensorType[SensorType["ROTATION"] = 5] = "ROTATION";
})(SensorType || (SensorType = {}));

function initSensorData(sensorType) {
  if (sensorType === SensorType.ROTATION) {
    return makeMutable({
      qw: 0,
      qx: 0,
      qy: 0,
      qz: 0,
      yaw: 0,
      pitch: 0,
      roll: 0
    });
  } else {
    return makeMutable({
      x: 0,
      y: 0,
      z: 0
    });
  }
}

export function useAnimatedSensor(sensorType, userConfig) {
  const ref = useRef({
    sensor: initSensorData(sensorType),
    unregister: () => {// NOOP
    },
    isAvailable: false,
    config: {
      interval: 0
    }
  });
  useEffect(() => {
    ref.current.config = {
      interval: 'auto',
      ...userConfig
    };
    const sensorData = ref.current.sensor;
    const id = registerSensor(sensorType, ref.current.config.interval === 'auto' ? -1 : ref.current.config.interval, data => {
      'worklet';

      sensorData.value = data;
    });

    if (id !== -1) {
      // if sensor is available
      ref.current.unregister = () => unregisterSensor(id);

      ref.current.isAvailable = true;
    } else {
      // if sensor is unavailable
      ref.current.unregister = () => {// NOOP
      };

      ref.current.isAvailable = false;
    }

    return () => {
      ref.current.unregister();
    };
  }, [sensorType, userConfig]);
  return ref.current;
}
